qiskit/dagcircuit/_dagcircuit.py

```diff
@@ -941,7 +941,7 @@ def substitute_node_with_dag(self, node, input_dag, wires=None):
         # Now that we know the connections, delete node
         self.multi_graph.remove_node(node)
         # Iterate over nodes of input_circuit
-        for m in input_dag.multi_graph:
+        for m in nx.topological_sort(input_dag.multi_graph):
```

Comment:
Can you add a comment explaining the nodes need to be sorted here (since it wasn't obvious from our first pass)? 

It looks like we update `full_pred_map` as we iterate over the DAG from referencing the `node_id`s of the `input_dag` to the `node_id`s of the current DAG, and so if all the predecessors of the current node haven't already been updated, we'll end up adding an incorrect edge (though I'm surprised we even get a run-able circuit at all, if this is the case.)

Reply:
I will have a deeper look. This PR is just to quickly fix so Aer can continue it business :)